### PR TITLE
fix: tell the planner a start time that survives its own thinking time

### PR DESCRIPTION
**Pull request by matthiasn**
Closes lotti3-y65. The first change from the eval epic aimed at *plan quality*
rather than at closing a hole — and the diagnosis turned out to be the
interesting part.

## The measurement

`compliedWithoutRejection` sat at 75% (glm-5.2) / 79% (qwen3.5-397b-a17b)
across the 48-cell matrix. One error dominated it, and on `lateStart` it was
total: **all 6/6 cells across both models** were rejected with
`blocks planned for today must not start before current time`.

## What it actually was

I assumed a prompt-clarity problem — the rule asks the model to notice
`<plan_date>` is today, find `<current_local_time>` (the last section in the
message), and derive a threshold. The judge bundle says otherwise:

```
now, as rendered in the prompt = 2026-07-26T15:00:00.005877
first block start              = 2026-07-26T15:00:00.000     ← rejected
```

Every rejected plan started the day at exactly 15:00 — the correct human
answer. **They lost by 5.877 milliseconds.**

The guard compares against `clock.now()` *when `draft_day_plan` executes*,
which is unavoidably later than the instant rendered into the prompt, because
the model thinks in between — 13s to 152s in this matrix. So any same-day plan
whose first block starts "now" is rejected, always. Complying would require the
model to predict its own inference latency.

This is a production race, not a prompt-quality issue, and it made the most
ordinary case there is — planning a day already underway — fail by
construction.

## The fix

**The guard is untouched and unweakened.** Only what the model is told to aim
at moved:

- `advertisedPlanningStart` — the first five-minute boundary at least three
  minutes after now — sits beside `earliestPlannableStart`, which the write
  path still enforces exactly as before.
- The drafting context carries it as `drafting.earliestStart`, present only on
  a same-day draft.
- The drafting rule points at that value instead of asking for a derivation.

`earliestPlannableStart` is now also the single source for both write-path call
sites, which previously inlined the same expression twice.

## Result

Re-ran `lateStart`, 6 cells, both models:

| | before | after |
| --- | ---: | ---: |
| past-start rejections | **6/6 cells** | **0/6** |
| `compliedWithoutRejection` | 0% | **100% / 100%** |

The cleanest single comparison is glm sample 3: same plan shape, same 15:05
start, same 17:00 end — but reached on the *first* attempt. Before, the model
was rejected at 15:00, retried, and arrived at 15:05 anyway. It was paying a
full round trip to discover a number we could simply have told it.

## What this cost, stated plainly

`withinWorkingHours` on that scenario went from 0 failures to 2 (glm, n=3).

The scenario offers 235 minutes of work in a ~115-minute window — **2x
oversubscribed** — so the model must drop or truncate, and it picks a different
victim each sample: truncate the migration (`respectsEstimates` fails, as it
did *before* this change) or overrun the day (`withinWorkingHours` fails, as it
does now). The instability is present in both runs; shaving 4% off an
impossible window exposed it rather than caused it. Filed as its own bug rather
than folded in here, since deciding what a correct oversubscribed plan looks
like also settles the open `respectsEstimates` question.

## Note on the rounding rule

My first implementation was wrong and the property test caught it: snapping to
the next five-minute boundary gives **one second** of headroom at 09:59:59,
which reproduces the exact bug. It now walks forward until the minimum headroom
holds, and the test asserts `advertised > enforced` at every minute of the hour
rather than at a couple of hand-picked instants.

## Verification

- `fvm dart analyze lib/features/daily_os_next test/features/daily_os_next` — clean
- `test/features/daily_os_next` — 2147 passing
- OKF validator — 89 concepts, 0 errors
- Live: 6 cells, 0 errors

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed Daily OS planning failures in edge cases where the requested plan started after the day’s first block had begun.
  * Improved timing validation so the planner uses a start time that remains valid when the plan arrives, reducing rejected first attempts and unnecessary retries.
* **Documentation**
  * Updated Daily OS capture and planning + wake prompt timing safeguards, clarifying “earliest start” behavior and when the planning window closes.
* **Tests**
  * Added coverage for earliest-start and planning-window guard behavior, including drafting and refine wake prompts.

### CODE REVIEW

File: lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart [hunk 1]
```diff
@@ -50,6 +50,71 @@ const plannedBlockStatesGuardedFromThePast = <PlannedBlockState>{
...
+    earliest.hour,
+    earliest.minute - (earliest.minute % step),
+  );
+  // Walk forward rather than snapping once: the nearest boundary can be a
+  // second away, which is no headroom at all.
+  var candidate = floor;
+  while (candidate.difference(earliest) < minimumPlanningHeadroom) {
+    candidate = candidate.add(advertisedStartGranularity);
+  }
+  return candidate;
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart:127]: Fixed, and you were right to call it Major — I had recreated the exact bug this PR fixes, from the other end of the day. Reproduced before changing anything:

    now=23:50 -> advertised=23:55   (fine)
    now=23:56 -> advertised=00:00 tomorrow
    now=23:58 -> advertised=00:05 tomorrow

`advertisedPlanningStart` now returns null once no five-minute slot remains inside the plan day.

On representing it distinctly — that turned out to matter more than the overflow itself. Null alone is ambiguous: it also means "future day, no constraint". A wake at 23:58 would then read as unconstrained and plan freely from this morning, and the guard would reject every block of it — strictly worse than the original bug. So `planningWindowClosed` reports the exhausted case separately, the drafting section carries it, and the rule tells the model to leave the day alone rather than add to it.

Regression tests as asked, plus a whole-day property test asserting nothing advertised ever escapes the plan day at any minute of any hour. 23:50 still advertises 23:55, because 23:55-00:00 is a legal block and closing early would silently drop the tail of the day.

File: lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart [hunk 2]
```diff
@@ -50,6 +50,71 @@ const plannedBlockStatesGuardedFromThePast = <PlannedBlockState>{
...
+    earliest.day,
+    earliest.hour,
+    earliest.minute - (earliest.minute % step),
+  );
+  // Walk forward rather than snapping once: the nearest boundary can be a
+  // second away, which is no headroom at all.
+  var candidate = floor;
+  while (candidate.difference(earliest) < minimumPlanningHeadroom) {
+    candidate = candidate.add(advertisedStartGranularity);
+  }
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart:114]: Fixed — same finding as CodeRabbit's, and your framing of it as "handle the exhausted planning window explicitly" is what the fix does. Details in that thread: `advertisedPlanningStart` returns null once no five-minute slot remains in the plan day, and `planningWindowClosed` distinguishes exhausted from unconstrained so a late wake cannot read as "plan freely".

File: lib/features/daily_os_next/agents/workflow/day_agent_prompt_builder.dart [hunk 3]
```diff
@@ -75,10 +75,13 @@ Capture matching rules:
 Drafting rules:
 - Every `ai` block passed to `draft_day_plan` must include a concrete reason.
 - Keep blocks inside the local plan day and within the user's capacity.
-- The user message includes a `<current_local_time>` section. When `<plan_date>`
-  is the same local day, do not plan any block that starts before that time.
-  Preserve already-started baseline blocks only when they represent existing
-  in-progress, completed, or dropped history.
+- When `drafting.earliestStart` is present, the day is already underway and no
+  new block may start before it. Use it as the start of your first block rather
+  than `<current_local_time>`: it is already rounded past the moment you are
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/workflow/day_agent_prompt_builder.dart]: Fixed. This was the more serious of the two: a straight regression, and I should have caught it when I replaced the rule.

The old rule was stated against `<current_local_time>` and applied to every wake. Keying the replacement on `drafting.earliestStart` silently narrowed it to drafting wakes only — while refine still enforces the identical guard through `proposePlanDiff`, and `draft_day_plan` is always exposed.

Both halves of your suggestion, since neither covers the whole surface alone:

- `RefineContext` now carries `earliestStart` as well, so refine gets the stated floor instead of deriving it — the same benefit drafting got, on a path that has the same guard.
- The rule keeps an explicit fallback for wakes carrying neither section, stated against `<current_local_time>` as before.

Tests pin both: a refine wake asserting the value reaches `refine.earliestStart`, and a bare wake asserting the fallback sentence survives in the system prompt.

File: lib/features/daily_os_next/agents/workflow/day_agent_prompt_builder.dart [hunk 4]
```diff
@@ -75,10 +75,17 @@ Capture matching rules:
...
-  Preserve already-started baseline blocks only when they represent existing
-  in-progress, completed, or dropped history.
+- When `drafting.earliestStart` or `refine.earliestStart` is present, the day is
+  already underway and no new or moved block may start before it. Use it rather
+  than `<current_local_time>`: it is already rounded past the moment you are
+  reading this, so a plan built on it survives your own thinking time, while one
+  built on the current instant is rejected for starting in the past. If neither
+  field is present and `<plan_date>` is the same local day as
+  `<current_local_time>`, the same rule still applies — do not start any block
+  before the current time. Preserve already-started baseline blocks only when
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/workflow/day_agent_prompt_builder.dart]: Fixed, and this one exposed the design error behind all three rounds — thank you for pushing on it.

You are right that the fallback was actively harmful: it pointed planning-day-only wakes at the raw `<current_local_time>`, which is the exact instant that gets rejected. I had written a "safety net" that instructed those wakes to do the broken thing.

The root cause was that I put the floor in the *mode contexts*. That gave me a list of wake types to keep in sync, and each review round found the next one I had missed — drafting, then refine, now scheduled. So it is now a single top-level `<planning_window>` section rendered on **every** wake, with the per-mode fields removed entirely. The constraint belongs to the day, not to how the wake was triggered.

Three deliberately distinct states, since collapsing any two misleads:

    {"earliestStart": …}  today, still plannable — start here
    {"closed": true}      today, no usable slot left — add nothing
    {}                    the day has not begun — nothing is in the past

A test now covers a wake carrying only `planning_day:<dayId>` and asserts it receives the padded value.

File: lib/features/daily_os_next/agents/workflow/day_agent_context_builder.dart [hunk 5]
```diff
@@ -752,6 +762,12 @@ extension DayAgentContextBuilder on DayAgentWorkflow {
       agentId: agentIdentity.agentId,
       dayId: wakeContext.dayId,
     );
-    return RefineContext(baselinePlan: baselinePlan);
+    return RefineContext(
+      baselinePlan: baselinePlan,
+      earliestStart: advertisedPlanningStart(
+        planDate: planDate,
+        now: clock.now(),
+      ),
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/workflow/day_agent_context_builder.dart]: Fixed by the same change — see the `day_agent_prompt_builder.dart` thread.

Rather than adding `planningWindowClosed` to `RefineContext` as well, I removed `earliestStart` from both mode contexts and made the whole thing one top-level `<planning_window>` section. Patching refine was what I did last round for the previous finding, and it just moved the hole along to scheduled wakes; a third patch would have moved it somewhere else. Refine now reads the same section as everything else, closed state included.

File: lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart [hunk 6]
```diff
@@ -50,6 +50,92 @@ const plannedBlockStatesGuardedFromThePast = <PlannedBlockState>{
...
+  // Late enough and the walk runs out of day: at 23:58 it lands on 00:05
+  // tomorrow, and `parsePlannedBlock` rejects anything outside the plan day —
+  // so advertising it would steer the model into the very rejection this
+  // function exists to prevent, at the other end of the day. Null means the
+  // window is closed, which [planningWindowClosed] distinguishes from the
+  // never-constrained future-day case.
+  final dayEnd = localDay(planDate).add(const Duration(days: 1));
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/daily_os_next/agents/service/day_agent_plan_parser.dart]: Fixed. `localDay(planDate).add(const Duration(days: 1))` is now `DateTime(day.year, day.month, day.day + 1)`.

Your two failure modes are exactly right, and both were reachable: spring-forward gives 01:00 tomorrow, so a late wake could advertise past midnight — reintroducing the bug this guard was added to prevent — and fall-back gives 23:00, closing the window an hour early and silently dropping the tail of the day.

The regression test asserts the observable behaviour rather than the arithmetic: the last usable slot (23:50 → 23:55) and the closing point (23:58 → closed) are identical on a spring-forward day, a fall-back day, and an ordinary control day.
